jdaviz/app.py

```diff
@@ -911,36 +913,84 @@ def add_data(self, data, data_label=None, notify_done=True):
+        check_unique : bool
+            Included so that this method can be used with data label retrieval in addition
+            to generation.
+
         Returns
         -------
-        A string data label used by Glue
+        data_label : str
+            A unique data label that at its root is either given by the user at load time
+            or created by jdaviz using a description of the loaded file's type.
+        """
+        data_label = None
+
+        if loaded_object is None:
+            pass
+        elif isinstance(loaded_object, str):
+            # This is either the full file path or the basename or the user input data label
+            if loaded_object.lower().endswith(('.jpg', '.jpeg', '.png')):
+                file_format = loaded_object.lower().split(".")[-1]
+                loaded_object = os.path.splitext(os.path.basename(loaded_object))[0]
+                data_label = f"{loaded_object}[{file_format}]"
+            elif not os.path.isfile(loaded_object) or re.search(r'(.+)(\[(.*?)\])$',
+                                                                loaded_object):
+                # Not a file path so assumed to be a data label
+                data_label = f"{loaded_object}"
+            else:
+                loaded_object = os.path.splitext(os.path.basename(loaded_object))[0]
+                data_label = f"{loaded_object}"
+        elif isinstance(loaded_object, fits.hdu.hdulist.HDUList):
+            if hasattr(loaded_object, 'file_name'):
+                data_label = f"{loaded_object.file_name}[HDU object]"
+            else:
+                data_label = "Unknown HDU object"
+        elif isinstance(loaded_object, Spectrum1D):
+            data_label = "Spectrum1D"
+        elif isinstance(loaded_object, NDData):
+            data_label = "NDData"
+        elif isinstance(loaded_object, np.ndarray):
+            data_label = "Array"
 
-        """
+        if data_label is None and alt_name is not None and isinstance(alt_name, str):
+            data_label = alt_name
+        elif data_label is None:
+            data_label = "Unknown"
 
-        # if path is not a file or already ends in [ ] extension, assume it is a data-label
-        if not os.path.isfile(path) or re.search(r'(.+)(\[(.*?)\])$', path):
-            return path
-        else:
-            assert ext, 'A data extension must be specified'
-            p = pathlib.Path(path)
-            stem = p.stem.split(os.extsep)[0]
-            label = f'{stem}[{ext}]'
-            return label
+        if ext is not None:
+            data_label = data_label + f"[{ext}]"
+
+        if check_unique:
+            data_label = self.return_unique_name(data_label)
+
+        return data_label
+
+    def return_unique_name(self, data_label):
+        if data_label is None:
+            data_label = "Unknown"
+
+        number_of_duplicates = 0
+        for label in self.data_collection.labels:
+            if data_label in label:
```

Comment:
How robust is this check?

Let's say we already loaded "M31" and "M32" (these are Messier object names). And then we load another unrelated image and be lazy and say `data_label="M"`. Will we detect false positive?

Reply:
It will, "M" will be loaded as "M (2)". How common would that use case be though? And how detrimental would it be for the user to see "M (2)" as opposed to "M"?

Reply:
It will be super confusing, so maybe ask the POs.

Reply:
Good call.

Reply:
I can see the workflow Pey Lian mentioned happening more often than not. Seeing "M (2)" would be confusing.

Reply:
Agree that this would be confusing. I like @kecnry 's suggestions.

Reply:
Thank you all for the feedback! The latest commit uses regex to determine if a data label already exists in the app. It took a while to get that working with all the different variations of data labels in our tests (names with brackets are now my least favorite), but this should now be ready for review!